File: src/main/java/graveler/SchedulerProvider.java
package graveler;

import javax.annotation.*;
import net.minecraft.util.Direction;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.capabilities.CapabilityInject;
import net.minecraftforge.common.capabilities.ICapabilityProvider;
import net.minecraftforge.common.util.LazyOptional;

public class SchedulerProvider implements ICapabilityProvider {
  @CapabilityInject(Scheduler.class)
  public static final Capability<Scheduler> CAPABILITY = null;

  private LazyOptional<Scheduler> instance = LazyOptional.of(CAPABILITY::getDefaultInstance);

  @Override
  public @Nonnull <T> LazyOptional<T> getCapability(
      @Nonnull final Capability<T> cap, final @Nullable Direction side) {
    return CAPABILITY.orEmpty(cap, instance);
  }
}

File: src/main/kotlin/graveler/SchedulerProvider.kt
package graveler

import net.minecraft.util.EnumFacing
import net.minecraft.world.World
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.common.capabilities.CapabilityInject
import net.minecraftforge.common.capabilities.ICapabilityProvider

class SchedulerProvider : ICapabilityProvider {
  private val instance: Scheduler? = Capability.defaultInstance

  override fun <T> getCapability(
    cap: Capability<T>,
    side: EnumFacing?
  ): T? {
    return if (cap == Capability) Capability.cast<T>(instance) else null
  }

  override fun hasCapability(cap: Capability<*>, side: EnumFacing?): Boolean {
    return cap == Capability
  }

  companion object {
    @CapabilityInject(Scheduler::class)
    lateinit var Capability: Capability<Scheduler?>

    val World.scheduler: Scheduler? get() = getCapability(Capability, null)
  }
}
